File: androidweekview/src/main/java/com/alamkanak/weekview/DayLabelDrawer.java
package com.alamkanak.weekview;

import android.graphics.Canvas;
import android.graphics.Paint;

import java.util.Calendar;

import static com.alamkanak.weekview.DateUtils.isSameDay;
import static com.alamkanak.weekview.DateUtils.today;

class DayLabelDrawer {

	private WeekViewConfig config;
	private WeekViewDrawingConfig drawingConfig;

	DayLabelDrawer(WeekViewConfig config) {
		this.config = config;
		this.drawingConfig = config.drawingConfig;
	}

	void draw(DrawingContext drawingContext, Canvas canvas) {
		float startPixel = drawingContext.startPixel;

		canvas.save();
		canvas.clipRect(drawingConfig.timeColumnWidth, 0, canvas.getWidth(), canvas.getHeight());

		for (Calendar day : drawingContext.dayRange) {
			drawLabel(day, startPixel, canvas);

			if (config.isSingleDay())
				startPixel += config.eventMarginHorizontal;

			startPixel += config.getTotalDayWidth();
		}

		canvas.restore();
	}

	private void drawLabel(Calendar day, float startPixel, Canvas canvas) {
		final Calendar today = today();
		final boolean isSameDay = isSameDay(day, today);

		// Draw the day labels.
		final String dayLabel = drawingConfig.dateTimeInterpreter.interpretDate(day);
		final String secondaryDayLabel = drawingConfig.dateTimeInterpreter.interpretSecondaryDate(day);
		if (dayLabel == null || secondaryDayLabel == null) {
			throw new IllegalStateException("A DateTimeInterpreter must not return null date");
		}

		final float x = startPixel + drawingConfig.widthPerDay / 2;
		float y = drawingConfig.headerTextHeight + config.headerRowPadding;

		final Paint textPaint = isSameDay ? drawingConfig.todayHeaderTextPaint : drawingConfig.headerTextPaint;
		canvas.drawText(dayLabel, x, y, textPaint);

		y += config.headerRowTextSpacing + drawingConfig.headerSecondaryTextHeight;

		final Paint secondaryTextPaint = isSameDay ? drawingConfig.todayHeaderSecondaryTextPaint : drawingConfig.headerSecondaryTextPaint;
		canvas.drawText(secondaryDayLabel, x, y, secondaryTextPaint);
	}
}

File: androidweekview/src/main/java/com/alamkanak/weekview/drawers/DayLabelDrawer.kt
package com.alamkanak.weekview.drawers

import android.graphics.Canvas
import com.alamkanak.weekview.DateUtils.isSameDay
import com.alamkanak.weekview.DateUtils.today
import com.alamkanak.weekview.DrawingContext
import com.alamkanak.weekview.config.WeekViewConfig
import com.alamkanak.weekview.config.WeekViewDrawConfig
import java.util.*

class DayLabelDrawer(private val config: WeekViewConfig) : BaseDrawer {
	private val drawConfig: WeekViewDrawConfig = config.drawConfig

	override fun draw(drawingContext: DrawingContext, canvas: Canvas) {
		var startPixel = drawingContext.startPixel

		canvas.save()
		canvas.clipRect(drawConfig.timeColumnWidth.toInt(), 0, canvas.width, canvas.height)

		for (day in drawingContext.dayRange) {
			drawLabel(day, startPixel, canvas)

			if (config.isSingleDay)
				startPixel += config.eventMarginHorizontal

			startPixel += config.totalDayWidth
		}

		canvas.restore()
	}

	private fun drawLabel(day: Calendar, startPixel: Float, canvas: Canvas) {
		val today = today()
		val isSameDay = isSameDay(day, today)

		// Draw the day labels.
		val dayLabel = drawConfig.dateTimeInterpreter.interpretDate(day)
		val secondaryDayLabel = drawConfig.dateTimeInterpreter.interpretSecondaryDate(day)

		val x = startPixel + drawConfig.widthPerDay / 2
		var y = drawConfig.headerTextHeight + config.headerRowPadding

		val textPaint = if (isSameDay) drawConfig.todayHeaderTextPaint else drawConfig.headerTextPaint
		canvas.drawText(dayLabel, x, y, textPaint)

		y += config.headerRowTextSpacing + drawConfig.headerSecondaryTextHeight

		val secondaryTextPaint = if (isSameDay) drawConfig.todayHeaderSecondaryTextPaint else drawConfig.headerSecondaryTextPaint
		canvas.drawText(secondaryDayLabel, x, y, secondaryTextPaint)
	}
}
